ftw/linkchecker/link_accumulation.py

```diff
@@ -61,64 +62,75 @@ def _separate_links_to_internal_external(self):
-                    self._append_to_link_and_relation_information_for_different_link_types(
-                        [[], uid, []], link_objs, obj)
-
-                if not isinstance(content, basestring):
-                    continue
-                # if there is a string having a valid scheme it will be embedded
-                # into a href, so we can use the same method as for the dexterity
-                # strings and do not need to change the main use case.
-                scheme = urlparse(content).scheme
-                if scheme and scheme in ['http', 'https']:
-                    content = 'href="%s"' % content
-                self._extract_and_append_link_objs(content, obj, link_objs)
-
+            link_objs = self._find_links_on_archetypes_brain_fields(obj, link_objs)
         else:
-            for name, field, schemata in self._iter_fields(obj.portal_type):
-                storage = schemata(obj)
-                fieldvalue = getattr(storage, name)
-                if not fieldvalue:
-                    continue
-
-                if IRelation.providedBy(field):
-                    if fieldvalue.isBroken():
-                        link = Link()
-                        link.complete_information_for_broken_relation_with_broken_relation_obj(
-                            obj, field)
-                        link_objs.append(link)
-
-                elif IURI.providedBy(field):
+            link_objs = self._find_links_on_dexterity_brain_fields(obj, link_objs)
+
+        return link_objs
+
+    def _find_links_on_dexterity_brain_fields(self, obj, link_objs):
+        for name, field, schemata in self._iter_fields(obj.portal_type):
+            storage = schemata(obj)
+            fieldvalue = getattr(storage, name)
+            if not fieldvalue:
+                continue
+
+            if IRelation.providedBy(field):
+                if fieldvalue.isBroken():
                     link = Link()
-                    link.complete_information_with_external_path(obj,
-                                                                 fieldvalue)
+                    link.complete_information_for_broken_relation_with_broken_relation_obj(
+                        obj, field)
                     link_objs.append(link)
 
-                elif IRichText.providedBy(field):
-                    content = fieldvalue.raw
-                    self._extract_and_append_link_objs(content, obj, link_objs)
+            elif IURI.providedBy(field):
+                link = Link()
+                link.complete_information_with_external_path(obj,
+                                                             fieldvalue)
+                link_objs.append(link)
+
+            elif IRichText.providedBy(field):
+                content = fieldvalue.raw
+                self._extract_and_append_link_objs(content, obj, link_objs)
+
+        return link_objs
+
+    def _find_links_on_archetypes_brain_fields(self, obj, link_objs):
```

Comment:
I know, it's just refactoring, but this method is quite long and has too many levels of indentation for my taste. Maybe you could consider to refactor this into two separate helper classes and a base class sharing some common code. The seeker could identify the type (AT/DX) and then ask the appropriate class to extract the links?

Reply:
In 84e23f8 I tried to split the LinkOnFieldSeeker further up. It instantiates either an ArchetypeSeeker or a DexteritySeeker which inherit from the BaseSeeker class (containing most of the methods used).
In a further commit, I will try to document more correct what they do.

I will look at the method mentioned. Maybe I'm going to do that as well later on.

Reply:
I tried to describe the task of every class in this module in b928a08. I hope the understanding was improved a little.